util/expvarx/expvarx.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package expvarx provides some extensions to the expvar package.
+package expvarx
+
+import (
+	"encoding/json"
+	"expvar"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"tailscale.com/types/lazy"
+)
+
+// SafeFunc is a wrapper around expvar.Func that guards against unbounded call
+// time and ensures that only a single call is in progress at any given time.
+type SafeFunc struct {
+	f      expvar.Func
+	limit  time.Duration
+	onSlow func(time.Duration, any)
+
+	mu       sync.Mutex
+	inflight *lazy.SyncValue[any]
+}
+
+// NewSafeFunc returns a new SafeFunc that wraps f. If f takes longer than limit
+// to execute then Value calls return nil. If non-nil, the parameter onSlow is a
+// function that is called when a slow operation completes, and is passed the
+// duration that f took to complete and the final value it produced.
+func NewSafeFunc(f expvar.Func, limit time.Duration, onSlow func(time.Duration, any)) *SafeFunc {
+	return &SafeFunc{f: f, limit: limit, onSlow: onSlow}
+}
+
+// Value acts similarly to expvar.Func.Value, but if the underlying function
+// takes longer than the configured limit, all callers will receive nil until
```

Comment:
Thoughts on panicking instead of `nil` ?

Reply:
That would create blind spots in monitoring again. I want data with holes rather than no data at all!